sdk/core/core-xml/src/xml.browser.ts

```diff
@@ -0,0 +1,165 @@
+
+const parser = new DOMParser();
+export function parseXML(str: string, opts?: { includeRoot?: boolean }): Promise<any> {
+  try {
+    const dom = parser.parseFromString(str, "application/xml");
+    throwIfError(dom);
+
+    let obj;
+    if (opts && opts.includeRoot) {
+      obj = domToObject(dom);
+    } else {
+      obj = domToObject(dom.childNodes[0]);
+    }
+
+    return Promise.resolve(obj);
+  } catch (err) {
+    return Promise.reject(err);
+  }
+}
+
+let errorNS = "";
+try {
+  errorNS = parser.parseFromString("INVALID", "text/xml").getElementsByTagName("parsererror")[0]
+    .namespaceURI!;
+} catch (ignored) {
+  // Most browsers will return a document containing <parsererror>, but IE will throw.
+}
+
+function throwIfError(dom: Document): void {
+  if (errorNS) {
+    const parserErrors = dom.getElementsByTagNameNS(errorNS, "parsererror");
+    if (parserErrors.length) {
+      throw new Error(parserErrors.item(0)!.innerHTML);
+    }
+  }
+}
+
+function isElement(node: Node): node is Element {
+  return !!(node as Element).attributes;
+}
+
+/**
+ * Get the Element-typed version of the provided Node if the provided node is an element with
+ * attributes. If it isn't, then undefined is returned.
+ */
+function asElementWithAttributes(node: Node): Element | undefined {
+  return isElement(node) && node.hasAttributes() ? node : undefined;
+}
+
+function domToObject(node: Node): any {
+  let result: any = {};
+
+  const childNodeCount: number = node.childNodes.length;
+
+  const firstChildNode: Node = node.childNodes[0];
+  const onlyChildTextValue: string | undefined =
+    (firstChildNode &&
+      childNodeCount === 1 &&
+      firstChildNode.nodeType === Node.TEXT_NODE &&
+      firstChildNode.nodeValue) ||
+    undefined;
+
+  const elementWithAttributes: Element | undefined = asElementWithAttributes(node);
+  if (elementWithAttributes) {
+    result["$"] = {};
+
+    for (let i = 0; i < elementWithAttributes.attributes.length; i++) {
+      const attr = elementWithAttributes.attributes[i];
+      result["$"][attr.nodeName] = attr.nodeValue;
+    }
+
+    if (onlyChildTextValue) {
+      result["_"] = onlyChildTextValue;
+    }
+  } else if (childNodeCount === 0) {
+    result = "";
+  } else if (onlyChildTextValue) {
+    result = onlyChildTextValue;
+  }
+
+  if (!onlyChildTextValue) {
+    for (let i = 0; i < childNodeCount; i++) {
+      const child = node.childNodes[i];
+      // Ignore leading/trailing whitespace nodes
+      if (child.nodeType !== Node.TEXT_NODE) {
+        const childObject: any = domToObject(child);
+        if (!result[child.nodeName]) {
+          result[child.nodeName] = childObject;
+        } else if (Array.isArray(result[child.nodeName])) {
+          result[child.nodeName].push(childObject);
+        } else {
+          result[child.nodeName] = [result[child.nodeName], childObject];
+        }
+      }
+    }
+  }
+
+  return result;
+}
+
+const serializer = new XMLSerializer();
```

Comment:
is there a particular reason why this is defined outside the scope of `stringifyXML`?

Reply:
I'm not sure, perhaps because we this gets used on every request, not reusing the serializer was expensive in terms of performance or working set?

Reply:
ah, makes sense.